src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -615,18 +617,31 @@ private static int readPort(File portFile) throws IOException {
         }
     }
 
+    /*
+     * Stops Jenkins and releases any system resources associated
+     * with it. If Jenkins is already stopped then invoking this
+     * method has no effect.
+     */
     public void stopJenkins() throws Throwable {
-        endpoint("exit").openStream().close();
-        if (!proc.waitFor(60, TimeUnit.SECONDS) ) {
-            System.err.println("Jenkins failed to stop within 60 seconds, attempting to kill the Jenkins process");
-            proc.destroyForcibly();
-            throw new AssertionError("Jenkins failed to terminate within 60 seconds");
+        if (!stopping) {
+            stopping = true;
```

Comment:
this is technically not threadsafe and could in some really narrow edge case still be call stop twice.

if you care about thread safety here then this should use an `AtomicBoolean` and `if (!stopping.getAndSet(true))` 

But more importantly what sets this back to `false`?

AFAICT this needs to be reset when starting Jenkins as you can have multiple start/stops of a `RealJenkinsRule` using 
multiple `then` inside a given test which will no longer close / stop the subsequent Jenkins processes.

This is very likely the cause of the test failures here.

Reply:
> AFAICT this needs to be reset when starting Jenkins as you can have multiple start/stops of a RealJenkinsRule using

+1

Reply:
Agree, AtomicBoolean is more reliable here, please see [55cc956](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/55cc9566d0a7ccf9bdf177f734c4751d3336bcb7)